bundles/org.eclipse.core.runtime/src/org/eclipse/core/runtime/Plugin.java

```diff
@@ -256,7 +258,12 @@ public final ILog getLog() {
 	 *  XXX Investigate the usage of a service factory (see also platform.getStateLocation)
 	 */
 	public final IPath getStateLocation() throws IllegalStateException {
-		return InternalPlatform.getDefault().getStateLocation(getBundle(), true);
+		if (stateLocation == null) {
+			// cache the value to avoid repeated java.io.File.mkdirs()
+			// does not matter if the value is computed twice in parallel
+			stateLocation = InternalPlatform.getDefault().getStateLocation(getBundle(), true);
```

Comment:
Maybe it would eb better to cache it at the InternalPlatform.getDefault().getStateLocation(...) side then so other calls can also benefit? Using a ConcurrentHashMap there would also avoid any double init as well...

Reply:
Such a map would need to be aware if bundles get uninstalled to clear out the bundle key (assuming bundles would be the key in such a map).  p2 doesn't support dynamic install/uninstall of bundles but we always tried to keep the core runtime aware that bundles can be installed/uninstalled dynamically while the platform is running.

Reply:
The call hierarchy of the internal methods is not so interesting/useful for optimization:

![image](https://user-images.githubusercontent.com/208716/161985770-0a4b7c23-e42b-4d90-894a-367f456a8ad8.png)

The simplicity of the proposal is nice and the performance avoids even a hash lookup...

Reply:
> p2 doesn't support dynamic install/uninstall of bundles

Actually, it partly and maybe even fully does. There used to be an "Apply" on the p2 install/uninstall dialog that was capable of refreshing the profile. The main issue isn't p2, but more some extension point consumers that are not reacting to extensions being installed/removed correctly; leading later on to unavailable features that were just installed, or ClassNotFound and other forms of inconsistencies

Reply:
> > p2 doesn't support dynamic install/uninstall of bundles
> 
> Actually, it partly and maybe even fully does. There used to be an "Apply" on the p2 install/uninstall dialog that was capable of refreshing the profile. The main issue isn't p2, but more some extension point consumers that are not reacting to extensions being installed/removed correctly; leading later on to unavailable features that were just installed, or ClassNotFound and other forms of inconsistencies

Agreed, I was oversimplifying the complete story.  The core platform and even p2 allows for dynamics (even the extension registry has events that can be listened to for dynamic updates). It is the rest of the system that isn't necessarily prepared for the complexity of dynamics.  In the end we thought it more simple to just have p2 default to force a restart on update.  The work to make such a dynamic system work across the complete community of plugins far outweighed the actual interest to have a running Eclipse instance update dynamically vs. just have it restart and work consistently.